### PR TITLE
Code optimize

### PR DESCRIPTION
**Pull request by daheige**
1.optimize gormSourceDir replace
2.fmt.Errorf adjust and Optimize for-break
3.avoid using the same name field and if..else optimization adjustment
4.optimization callbacks/create.go Create func if...else logic
5.fix FileWithLineNum func and add gormSourceDir unit test

### CODE REVIEW

File: schema/field.go [hunk 1]
```diff
@@ -198,36 +198,36 @@ func (schema *Schema) ParseField(fieldStruct reflect.StructField) *Field {
...
+				schema.err = fmt.Errorf("failed to parse %s as default value for float, got error: %v", field.DefaultValue, err)
 			}
 		}
 	case reflect.String:
 		field.DataType = String
 
 		if field.HasDefaultValue && !skipParseDefaultValue {
 			field.DefaultValue = strings.Trim(field.DefaultValue, "'")
-			field.DefaultValue = strings.Trim(field.DefaultValue, "\"")
+			field.DefaultValue = strings.Trim(field.DefaultValue, `\"`)
```
Comment (jinzhu) [inline schema/field.go:230]: \ is not necessary?

Comment (daheige) [reply schema/field.go:230]: ok

File: utils/utils.go [hunk 2]
```diff
@@ -15,7 +14,7 @@ var gormSourceDir string
 
 func init() {
 	_, file, _, _ := runtime.Caller(0)
-	gormSourceDir = regexp.MustCompile(`utils.utils\.go`).ReplaceAllString(file, "")
```
Comment (jinzhu) [inline utils/utils.go:18]: this is used to remove `utils/utils.go` or `utils\utils.go`

**Comment by daheige** [comment]
Look at your regular replacement implementation code logic here. In fact, it's utils.utils.go in the replacement path. There's no need to use the regular replacement mode here, just use the string replacement method.


------------------ Original ------------------
From: Jinzhu ***@***.***&gt;
Date: Mon,May 31,2021 2:23 PM
To: go-gorm/gorm ***@***.***&gt;
Cc: heige ***@***.***&gt;, Author ***@***.***&gt;
Subject: Re: [go-gorm/gorm] Code optimize (#4415)





 
@jinzhu commented on this pull request.
 
 
In utils/utils.go:
 &gt; @@ -15,7 +14,7 @@ var gormSourceDir string    func init() {  _, file, _, _ := runtime.Caller(0) -gormSourceDir = regexp.MustCompile(`utils.utils\.go`).ReplaceAllString(file, "")  
this is used to remove utils/utils.go or utils\utils.go

**Comment by jinzhu** [comment]
not really, you can write a test to check it.

**Comment by daheige** [comment]
As you said, here is the path to get gormSourceDir. It is recommended that you use the filepath mode to get it, because it is compatible with various operating systems and does not need to replace utils/utils.go or utils/utils.go. For specific unit tests, please see this Branch
https://github.com/daheige/gorm/commit/4cae7162e1ed80843c6b920fe178343fa9595802

**Comment by daheige** [comment]
> not really, you can write a test to check it.

As you said, here is the path to get gormSourceDir. It is recommended that you use the filepath mode to get it, because it is compatible with various operating systems and does not need to replace utils/utils.go or utils/utils.go. For specific unit tests, please see this Branch
https://github.com/daheige/gorm/commit/4cae7162e1ed80843c6b920fe178343fa9595802

File: utils/utils.go [hunk 3]
```diff
@@ -15,17 +15,22 @@ var gormSourceDir string
...
+	gormSourceDir = filepath.Dir(filepath.Dir(file))
 }
 
+// FileWithLineNum return the file name and line number of the current file
 func FileWithLineNum() string {
-	for i := 2; i < 15; i++ {
+	for i := 1; i < 15; i++ {
 		_, file, line, ok := runtime.Caller(i)
 
+		// fmt.Println("file:", file, " line: ", line, "ok: ", ok)
```
Comment (jinzhu) [inline utils/utils.go:28]: please remove the debug code.

File: utils_filenum_test.go [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package gorm
```
Comment (jinzhu) [inline utils_filenum_test.go:1]: the tests don't make sense.

**Comment by daheige** [comment]
@jinzhu i see. I mainly test the correctness of the function of getting the number of file lines that you wrote and I wrote here. Later, I will remove the meaningless test file that you said.


------------------ Original ------------------
From: Jinzhu ***@***.***&gt;
Date: Mon,Jun 7,2021 10:33 AM
To: go-gorm/gorm ***@***.***&gt;
Cc: heige ***@***.***&gt;, Author ***@***.***&gt;
Subject: Re: [go-gorm/gorm] Code optimize (#4415)





 
@jinzhu commented on this pull request.
 
 
In utils_filenum_test.go:
 &gt; @@ -0,0 +1,12 @@ +package gorm  
the tests don't make sense.

**Comment by daheige** [comment]
ok，thanks ,i will remove it


------------------ Original ------------------
From: Jinzhu ***@***.***&gt;
Date: Mon,Jun 7,2021 10:33 AM
To: go-gorm/gorm ***@***.***&gt;
Cc: heige ***@***.***&gt;, Author ***@***.***&gt;
Subject: Re: [go-gorm/gorm] Code optimize (#4415)





 
@jinzhu commented on this pull request.
 
 
In utils/utils.go:
 &gt;  _, file, line, ok := runtime.Caller(i)   +// fmt.Println("file:", file, " line: ", line, "ok: ", ok)  
please remove the debug code.

**Comment by daheige** [comment]
@jinzhu i has removed them.

**Comment by Kisesy** [comment]
@daheige 
gormSourceDir 那里不能这样写：
> gormSourceDir = filepath.Dir(filepath.Dir(file))

因为 `runtime.Caller(0)` 返回的路径分隔符固定是斜杠: `"/"`，在Windows上也是斜杠，你用 filepath.Dir 来处理是不行的
输出日志就会是：
D:/dev/Go/bin/pkg/mod/gorm.io/gorm@v1.21.11-0.20210610022128-50e85e14d451/logger/logger.go:169

简单的解决办法就是把 filepath.Dir() 换成  path.Dir()

**Comment by daheige** [comment]
What you said is wrong. Filepath is cross platform. You can learn about it. Thank you.




------------------ Original ------------------
From: kisesy ***@***.***&gt;
Date: Thu,Jun 10,2021 4:33 PM
To: go-gorm/gorm ***@***.***&gt;
Cc: heige ***@***.***&gt;, Mention ***@***.***&gt;
Subject: Re: [go-gorm/gorm] Code optimize (#4415)





 
@daheige
 gormSourceDir 那里不能这样写：
  
gormSourceDir = filepath.Dir(filepath.Dir(file))
  
因为 runtime.Caller(0) 返回的路径分隔符固定是斜杠: "/"，在Windows上也是斜杠，你用 filepath.Dir 来处理是不行的
 输出日志就会是：
 ***@***.***/logger/logger.go:169
 
简单的解决办法就是把 filepath.Dir() 换成  path.Dir()

**Comment by Kisesy** [comment]
@daheige 
不好意思, 我上面没有说明白，filepath 是可以处理多平台的，但是它返回的结果却是本系统的规则
比如在Windows上 `filepath.Dir("/a/b/c")` -> `\a\b`，这里返回的结果是反斜杠，也是 Windows 的分隔符

如果用这个结果来跑下面的代码
```go
_, file, line, ok := runtime.Caller(i)
// 下面是代码执行时变量的值
// file: D:/dev/Go/bin/pkg/mod/gorm.io/gorm@v1.21.11-0.20210610022128-50e85e14d451/logger/logger.go
// gormSourceDir: D:\dev\Go\bin\pkg\mod\gorm.io\gorm@v1.21.11-0.20210610022128-50e85e14d451
if ok && (!strings.HasPrefix(file, gormSourceDir) || strings.HasSuffix(file, "_test.go")) {
	return file + ":" + strconv.FormatInt(int64(line), 10)
}
```
我之前说了 runtime.Caller() 返回的是斜杠，上面 gormSourceDir 用 filepath.Dir() 处理完的是反斜杠，
那么 strings.HasPrefix() 可不是跨平台的，一个是斜杠一个是反斜杠，这里永远是不会判断为真的

**Comment by daheige** [comment]
!strings.HasPrefix(file, gormSourceDir) || strings.HasSuffix(file, "_test.go")
I generally know that there may have been a problem before you mentioned here, but here you can use filepath to get the path of this file, and then judge whether it is included or not. This may be a problem left over by history. Let me confirm that if there is a problem, you can correct it. As for the runtime. Caller return /, you can write a test case. Thank you again for some of the points you mentioned.


------------------ Original ------------------
From: kisesy ***@***.***&gt;
Date: Fri,Jun 11,2021 2:35 PM
To: go-gorm/gorm ***@***.***&gt;
Cc: heige ***@***.***&gt;, Mention ***@***.***&gt;
Subject: Re: [go-gorm/gorm] Code optimize (#4415)





 
@daheige
 不好意思, 我上面没有说明白，filepath 是可以处理多平台的，但是它返回的结果却是本系统的规则
 比如在Windows上 filepath.Dir("/a/b/c") -&gt; \a\b，这里返回的结果是反斜杠，也是 Windows 的分隔符
 
如果用这个结果来跑下面的代码
 _, file, line, ok := runtime.Caller(i) // 下面是代码执行时变量的值 // file: ***@***.***/logger/logger.go // gormSourceDir: ***@***.*** if ok &amp;&amp; (!strings.HasPrefix(file, gormSourceDir) || strings.HasSuffix(file, "_test.go")) { return file + ":" + strconv.FormatInt(int64(line), 10) }
 
我之前说了 runtime.Caller() 返回的是斜杠，上面 gormSourceDir 用 filepath.Dir() 处理完的是反斜杠，
 那么 strings.HasPrefix() 可不是跨平台的，一个是斜杠一个是反斜杠，这里永远是不会判断为真的

**Comment by daheige** [comment]
@kisesy thanks.
!strings.HasPrefix(filepath.Dir(file), gormSourceDir) || strings.HasSuffix(file, "_test.go")
Thank you for your suggestion. I'll fix it today.


------------------ Original ------------------
From: kisesy ***@***.***&gt;
Date: Fri,Jun 11,2021 2:35 PM
To: go-gorm/gorm ***@***.***&gt;
Cc: heige ***@***.***&gt;, Mention ***@***.***&gt;
Subject: Re: [go-gorm/gorm] Code optimize (#4415)





 
@daheige
 不好意思, 我上面没有说明白，filepath 是可以处理多平台的，但是它返回的结果却是本系统的规则
 比如在Windows上 filepath.Dir("/a/b/c") -&gt; \a\b，这里返回的结果是反斜杠，也是 Windows 的分隔符
 
如果用这个结果来跑下面的代码
 _, file, line, ok := runtime.Caller(i) // 下面是代码执行时变量的值 // file: ***@***.***/logger/logger.go // gormSourceDir: ***@***.*** if ok &amp;&amp; (!strings.HasPrefix(file, gormSourceDir) || strings.HasSuffix(file, "_test.go")) { return file + ":" + strconv.FormatInt(int64(line), 10) }
 
我之前说了 runtime.Caller() 返回的是斜杠，上面 gormSourceDir 用 filepath.Dir() 处理完的是反斜杠，
 那么 strings.HasPrefix() 可不是跨平台的，一个是斜杠一个是反斜杠，这里永远是不会判断为真的

**Comment by Kisesy** [comment]
@daheige 
只需把 filepath 改成 path，比如：`gormSourceDir = path.Dir(path.Dir(file))`，因为 `runtime.Caller()` 无论在什么平台都会返回斜杠

**Comment by daheige** [comment]
Maybe you're right. We need to consider the cross platform of directory.


------------------ Original ------------------
From: kisesy ***@***.***&gt;
Date: Fri,Jun 11,2021 5:45 PM
To: go-gorm/gorm ***@***.***&gt;
Cc: heige ***@***.***&gt;, Mention ***@***.***&gt;
Subject: Re: [go-gorm/gorm] Code optimize (#4415)





 
@daheige
 只需把 filepath 改成 path，比如：gormSourceDir = path.Dir(path.Dir(file))，因为 runtime.Caller() 无论在什么平台都会返回斜杠

**Comment by Kisesy** [comment]
@daheige 我没法测试，但这其实是个遗留问题
https://github.com/golang/go/issues/18151
https://github.com/golang/go/issues/3335

**Comment by daheige** [comment]
@kisesy thank you very much!
https://github.com/go-gorm/gorm/pull/4456




 原始邮件 
发件人: ***@***.***>
收件人: ***@***.***>
抄送: ***@***.***>; ***@***.***>
发送时间: 2021年6月11日(周五) 18:14
主题: Re: [go-gorm/gorm] Code optimize (#4415)


@daheige 我没法测试，但这其实是个遗留问题
 golang/go#18151
 golang/go#3335

**Pull request merged**